src/inlineExpression.tsx

```diff
@@ -0,0 +1,127 @@
+import React, { useRef, useEffect, useMemo } from 'react';
+import { useCellMetadata } from './cellMetadataProvider';
+import { SingletonLayout, Widget } from '@lumino/widgets';
+import { IRenderMimeRegistry } from '@jupyterlab/rendermime';
+import { IExpressionResult, isOutput } from './userExpressions';
+
+export interface IRenderedExpressionOptions {
+  expression: string;
+  trusted: boolean;
+  rendermime: IRenderMimeRegistry;
+  safe?: 'ensure' | 'prefer' | 'any';
+}
+
+export class RenderedExpressionError extends Widget {
+  constructor() {
+    super();
+    this.addClass('im-RenderedExpressionError');
+  }
+}
+
+export class RenderedExpression extends Widget {
+  readonly expression: string;
+  readonly trusted: boolean;
+  readonly rendermime: IRenderMimeRegistry;
+  readonly safe?: 'ensure' | 'prefer' | 'any';
+
+  constructor(options: IRenderedExpressionOptions) {
+    super();
+
+    this.trusted = options.trusted;
+    this.expression = options.expression;
+    this.rendermime = options.rendermime;
+    this.safe = options.safe;
+
+    this.addClass('im-RenderedExpression');
+
+    // We can only hold one renderer at a time
+    const layout = (this.layout = new SingletonLayout());
+    layout.widget = new RenderedExpressionError();
+  }
+
+  renderExpression(payload: IExpressionResult): Promise<void> {
+    const layout = this.layout as SingletonLayout;
+
+    let options: any;
+    if (isOutput(payload)) {
+      // Output results are simple to reinterpret
+      options = {
+        trusted: this.trusted,
+        data: payload.data,
+        metadata: payload.metadata
+      };
+    } else {
+      // Errors need to be formatted as stderr objects
+      options = {
+        data: {
+          'application/vnd.jupyter.stderr':
+            payload.traceback.join('\n') ||
+            `${payload.ename}: ${payload.evalue}`
+        }
+      };
+    }
+
+    // Invoke MIME renderer
+    const model = this.rendermime.createModel(options);
+
+    // Select preferred mimetype for bundle
+    const mimeType = this.rendermime.preferredMimeType(model.data, this.safe);
+    if (mimeType === undefined) {
+      console.error("Couldn't find mimetype for ", model);
+
+      // Create error
+      layout.widget = new RenderedExpressionError();
+      return Promise.resolve();
+    }
+
+    // Create renderer
+    const renderer = this.rendermime.createRenderer(mimeType);
+    layout.widget = renderer;
+    console.assert(renderer.isAttached, 'renderer was not attached!', renderer);
+    // Render model
+    return renderer.renderModel(model);
+  }
+}
+
+export function InlineRenderer({ value }: { value?: string }): JSX.Element {
```

Comment:
This react component gets attached in the `inlineExpression` for mdast. We have to be careful here to create the rendermime one time, and then update it when the expression result changes. There is also ways to dispose of the renderer when this component gets removed.

Reply:
@stevejpurves you might have comments here from your experience working with react and thebe?!

Reply:
this is ok especially if it works. I have has issues with render passes previously before a ref was resolved and solved that by using "callback refs" to only attach to the dom once the ref was available. I think that is more of an issue in components with conditional rendering -- which this one has i.e. when there is no `expressionResult` but if this pattern works and we reliably get the output attached to the dom, then no need to change it.